server/src/main/java/org/elasticsearch/script/TimeSeriesCounter.java

```diff
@@ -0,0 +1,399 @@
+         *        [_]        [_]         [2]          [3]        [4]
+         *  |___________|___________|___________|______1____|___________|
+         *                         240[e]->    260->       280->       299
+         *
+         * Present
+         *       [0]         [1][b]      [2][g]      [3]         [4]
+         *  |_____1_____|_____1_____|___________|___________|___________|
+         * 300[a]->    320->       340[d]->    360->       380->       399
+         *
+         * Future
+         *       [0]         [_]         [_]         [_]         [_]
+         *  |_____0_____|___________|___________|___________|___________|
+         * 400[c]->    420[f]->
+         *
+         */
+        protected final long resolution;
+        protected final long duration;
+        protected final long[] buckets;
+
+        // The start time of buckets[0]. bucket(t + (i * duration)) is the same for all i. startOfEpoch
+        protected long startOfEpoch;
+        protected int curBucket = 0;
+
+        /**
+         * Create a Counter that covers duration seconds at the given resolution.  Duration must be divisible by resolution.
+         */
+        public Counter(long resolution, long duration) {
+            if (resolution <= 0) {
+                throw new IllegalArgumentException("resolution [" + resolution + "] must be greater than zero");
+            } else if (duration <= 0) {
+                throw new IllegalArgumentException("duration [" + duration + "] must be greater than zero");
+            } else if (duration % resolution != 0) {
+                throw new IllegalArgumentException("duration [" + duration + "] must divisible by resolution [" + resolution + "]");
+            }
+            this.resolution = resolution;
+            this.duration = duration;
+            this.buckets = new long[(int) (duration / resolution)];
+            assert buckets.length > 0;
+        }
+
+        /**
+         * Increment the counter at time {@code t}, expressed in seconds.
+         */
+        public void inc(long t) {
+            if (t < nextBucketStartTime()) {
+                buckets[curBucket]++;
+            } else if (t >= counterExpired()) {
+                reset(t);
+            } else {
+                int dstBucket = bucket(t);
+                for (int i = nextBucket(curBucket); i != dstBucket; i = nextBucket(i)) {
+                    buckets[i] = 0;
+                }
+                curBucket = dstBucket;
+                buckets[curBucket] = 1;
+                if (t >= nextEpoch()) {
+                    startOfEpoch = epoch(t);
+                }
+            }
+        }
+
+        /**
+         * sum for the duration of the counter until {@code end}.
+         */
+        public long sum(long end) {
+            long start = end - duration;
+            if (start >= nextBucketStartTime() || start < 0) {
+                return 0;
+            }
+
+            if (start < earliestTimeInCounter()) {
```

Comment:
This is to avoid handling negative values, as the math gets weird when time goes negative.  We're expecting `System.currentTimeMillis()` via `ThreadPool` and `TimeSeriesCounter.now()` is ensures time to be non-negative are enforce this requirement.